Add support for retrieving multiple results.

petermattis opened the pull request:
Reworked how the msg-ready command (`Z`) is processed. Previously
execution of a query would look for the msg-ready command before
completing the operation. Now, when executing a query, the driver places
the connection into a state where it knows there may be more results. If
another query is subsequently executed, the driver waits for the
msg-ready command to arrive, discarding any other commands, before
sending the new query. But if the special `NEXT` query is executed, the
driver looks for another set of results on the connection. If no such
results are found, ErrNoMoreResults is returned.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pq/1)

<!-- Reviewable:end -->

petermattis committed:
Properly support multiple results.

Reworked how the msg-ready command (`Z`) is processed. Previously
execution of a query would look for the msg-ready command before
completing the operation. Now, when executing a query, the driver places
the connection into a state where it knows there may be more results. If
another query is subsequently executed, the driver waits for the
msg-ready command to arrive, discarding any other commands, before
sending the new query.

tamird commented:
Connection pinning concerns still stand. I'd really like to see this functionality wrapped up into a function that takes care of all the shenanigans.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[conn.go, line 624 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bZ5aQc2HKVkRDSrM8:-K9bZ5aQc2HKVkRDSrM9:-1165217051) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L624)):_
along with `ErrNoMoreResults`, this should be an exported constant

---

_[conn.go, line 638 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bZZGD4985zNzinqPb:-K9bZZGD4985zNzinqPc:1036922401) ([raw file](https://github.com/cockroachdb/pq/blob/8ad2b298cadd691a77015666a5372eae5dbfac8f/conn.go#L638)):_
what's this change about? looks like unrelated cleanup - separate commit or back it out - we don't want to diverge more than necessary.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9b_PcdDu1GX090dgL4:-65271356) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
what's happening here?

---

_[conn.go, line 1641 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_ActpKbfwhwI4sSR:-K9b_ActpKbfwhwI4sSS:-133591105) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1641)):_
the lack of a default here is worrying

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1#-:-K9ba2WdM_x8qIsZKl6i:1850764157)_

<!-- Sent from Reviewable.io -->

tbg commented on conn.go:
```diff
@@ -200,7 +204,7 @@ func (c *conn) handlePgpass(o values) {
 			}
 		}
 		return append(fs, string(f))
-	}	
```
if you're upstreaming this (or trying to) it probably makes sense to factor out the drive-by changes in a separate commit.

tbg commented on conn_test.go:
```diff
@@ -339,6 +340,56 @@ func TestRowsCloseBeforeDone(t *testing.T) {
+	if err := db.QueryRow("NEXT").Scan(&val); err != ErrNoMoreResults {
+		t.Fatalf("expected %s, but found %v", ErrNoMoreResults, err)
+	}
+}
```
what about a test of discarding the extra results?

tbg commented on conn.go:
```diff
@@ -1436,7 +1467,11 @@ func (rs *rows) Next(dest []driver.Value) (err error) {
+			if err != nil {
+				return err
+			}
+			return io.EOF
```
what's happening here?

petermattis committed:
Remove unnecessary allocation.

tbg commented on conn.go:
```diff
@@ -614,56 +620,73 @@ func (cn *conn) simpleExec(q string) (res driver.Result, commandTag string, err
-	b.string(q)
-	cn.send(b)
+	querySent := false
+	nextResult := q == "NEXT"
```
pull out magic constant into named const variable?

petermattis commented:
Moved the drive-by change to a separate commit.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[conn.go, line 624 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bZ5aQc2HKVkRDSrM8:-K9bbELfggRHPQULIdTO:-499146009) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L624)):_
Done.

---

_[conn.go, line 638 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bZZGD4985zNzinqPb:-K9baZGk21rwg2q1_mjp:2128027985) ([raw file](https://github.com/cockroachdb/pq/blob/8ad2b298cadd691a77015666a5372eae5dbfac8f/conn.go#L638)):_
Yes, just unrelated cleanup of unused structure. I'm going to push this upstream (or try to).

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bajlYhNcri-AIVvrC:867911346) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
This is the correct way to process a result. A `C` or `I` command indicates the end of data rows. The previous behavior of looking for a `Z` was just completely incorrect.

---

_[conn.go, line 1641 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_ActpKbfwhwI4sSR:-K9batrpulqkBjdKceDn:-99243572) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1641)):_
As far as I can tell from the postgres docs, this is what is expected. The client "syncs" the connection by waiting for a `Z`.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1#-:-K9bbjbKErdxcNi4fJom:-1575739757)_

<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bd2z1_-v2JlLAlsg4:958622006) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
I see. That also seems like a standalone change that deserves its own commit.

---

_[conn.go, line 1641 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_ActpKbfwhwI4sSR:-K9bdAnn-nvEL1q5ftmP:547841965) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1641)):_
and it's OK to discard bytes until the Z arrives?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[conn.go, line 624 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bXq5pzbX8Km6iZ_G1-r1-624:-K9bdLjqxHQXHFov_t3y:-499146009) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L624)):_
Done.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bdfUQscLrykp5pQts:-900829963) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
What is the value in moving this to its own commit? That just feels like busy work for the author of the code. And note that this isn't easily separable. We'd still need all of the `readyForQuery` stuff. 

---

_[conn.go, line 1641 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_ActpKbfwhwI4sSR:-K9be2IkQcR4kCMKT1EM:677251276) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1641)):_
Yes.

---

_[conn_test.go, line 363 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bXq5pzbX8Km6iZ_G2-r1-363:-K9bddHgOevUFbu-6Zj4:-499146009) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn_test.go#L363)):_
Done.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

petermattis commented:
@tamird Do you have a suggestion for how to wrap up this functionality into a function to take care of the shenanigans?

---

Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1#-:-K9bejsnwj5JLDPGkaGP:-1749223389)_

<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bejY9KmRqEfjcfhSf:-1772700151) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
I'm missing something, because if the current code is just wrong, then this change should be separable.

I don't think this change will be accepted upstream in a single PR, which is why I think it should be separated.

---

_[conn.go, line 1641 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_ActpKbfwhwI4sSR:-K9bfNeL_0e7SUaZKSr6:-579060256) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1641)):_
Can you add a comment that points to docs? That's quite surprising given the apparent strictness of similar code in this file.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bfv8BC7ZeJ-mDNYsh:1675074421) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
The 5 lines of change are fix a bug when a query contains multiple statements, but they require all of the `readyForQuery` changes as well.

But let's focus on whether the code is correct first and worry about how to get it pushed upstream second. It looks very likely that the handling of multiple results will not be accepted upstream. That's fine, that's what forks are for.

---

_[conn.go, line 1641 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_ActpKbfwhwI4sSR:-K9bgVgzfOAPxjuI133N:-402878313) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1641)):_
`Since a query string could contain several queries (separated by semicolons), there might be several such response sequences before the backend finishes processing the query string. ReadyForQuery is issued when the entire string has been processed and the backend is ready to accept a new query string.`

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

tbg commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9biFZ9mCgUwbnrPTuy:1577864643) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_

> What is the value in moving this to its own commit?

Maybe no value doing it now, but had it been in its own commit from the start, you would've explained this in your commit message. Instead, Tamir and I independently tried to understand this change without explanation, which costs time.

---

_[conn.go, line 1339 [r3]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bi3NbXgTajsy2Ir9M:-K9bi3NbXgTajsy2Ir9N:-1572509625) ([raw file](https://github.com/cockroachdb/pq/blob/943e5752d26737e49bac2dbf1d8fed446cd36ce0/conn.go#L1339)):_
s/has/as/

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bn032RrHjU_1byK7U:1487480231) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
I can't tell if this code is correct without this change separated out. With this change (but without support for NEXT) do the tests pass?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

petermattis committed:
Add support for retrieving multiple results.

If the special `NEXT` query is executed, the driver looks for another
set of results on the connection. If no such results are found,
ErrNoMoreResults is returned.

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[conn.go, line 1470 [r1]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9b_PcdDu1GX090dgL3:-K9bowFc93QpisdPbWQf:-484112061) ([raw file](https://github.com/cockroachdb/pq/blob/8b0d5d6ee14efc79956d3ff51c72cacd23a56752/conn.go#L1470)):_
Yes. The support for `NEXT` is purely in `conn.SimpleQuery`. Refactored the commits to increase reviewer happiness.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

tbg commented:
Reviewed 1 of 1 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1)_

<!-- Sent from Reviewable.io -->

petermattis commented:
Anything else to do here?

---

Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

_[conn.go, line 1339 [r3]](https://reviewable.io:443/reviews/cockroachdb/pq/1#-K9bi3NbXgTajsy2Ir9M:-K9c3x6xtkUswOZuaz-Y:-499146009) ([raw file](https://github.com/cockroachdb/pq/blob/943e5752d26737e49bac2dbf1d8fed446cd36ce0/conn.go#L1339)):_
Done.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1#-:-K9c3zD8bkmN8GOzgOW0:-2136272271)_

<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/pq/1#-:-K9cXvmMnLz5jqFCWqBp:676834233)_

<!-- Sent from Reviewable.io -->
